diff -up net-snmp-5.7.2/agent/mibgroup/hardware/fsys/fsys_mntent.c.orig net-snmp-5.7.2/agent/mibgroup/hardware/fsys/fsys_mntent.c
--- net-snmp-5.7.2/agent/mibgroup/hardware/fsys/fsys_mntent.c.orig	2018-04-05 18:29:29.042682282 +0200
+++ net-snmp-5.7.2/agent/mibgroup/hardware/fsys/fsys_mntent.c	2018-04-05 19:03:16.270523986 +0200
@@ -142,6 +142,7 @@ _fsys_type( char *typename )
               !strcmp(typename, MNTTYPE_SIMFS) ||
               !strcmp(typename, MNTTYPE_VZFS) ||
               !strcmp(typename, MNTTYPE_BTRFS) ||
+              !strcmp(typename, MNTTYPE_ACFS) ||
               !strcmp(typename, MNTTYPE_LOFS))
        return NETSNMP_FS_TYPE_OTHER;
 
diff -up net-snmp-5.7.2/agent/mibgroup/hardware/fsys/mnttypes.h.orig net-snmp-5.7.2/agent/mibgroup/hardware/fsys/mnttypes.h
--- net-snmp-5.7.2/agent/mibgroup/hardware/fsys/mnttypes.h.orig	2018-04-05 18:28:36.694091896 +0200
+++ net-snmp-5.7.2/agent/mibgroup/hardware/fsys/mnttypes.h	2018-04-05 20:14:55.695813461 +0200
@@ -151,6 +151,9 @@
 #ifndef MNTTYPE_BTRFS
 #define MNTTYPE_BTRFS    "btrfs"
 #endif
+#ifndef MNTTYPE_ACFS
+#define MNTTYPE_ACFS  "acfs"
+#endif
 
 /*
  *  File systems to skip
